docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,136 @@
+
+## Example A: Random Number Construct `rand()`
+
+The random number construct `rand()` is a language construct. It is part of the syntax of the PL:
+
+```
+Stmt ::= "rand" "(" ")"
+```
+
+### Axioms
+
+The intended semantics of `rand()` is that it can rewrite to every integers. Here is its axiom
+
+```
+(RAND) \forall I:Int . rand() -> * I
+// equivalent form(s)
+//     rand() -> \forall I:Int . * I
+//     rand() -> * I
+// non-equivalent form(s)
+//     rand() -> * (\forall I:Int . I)
+```
+
+Note that `(RAND)` axiom is logically equivalent to the following infinitely many axioms
+
+```
+(RAND0)  rand() -> * 0
+(RAND1)  rand() -> * 1
+(RAND2)  rand() -> * 2
+  ...    ...
+(RAND-1) rand() -> * (-1)
+(RAND-2) rand() -> * (-2)
+  ...    ...
+```
+
+It is not easy to define many of the variants of `rand()`. For example. `randPrime()` can rewrite to every _prime numbers_. Here is its axiom
+
+```
+(RANDPRIME) \forall P:Int . isPrime(P) -> (randPrime() -> * P)
+// equivalent form(s)
+//          \forall P:Int . (randPrime() /\ isPrime(P)) -> *P
+//          \forall P:Int . randPrime() -> (isPrime(P) -> * P)
+//          randPrime() -> \forall P:Int . (isPrime(P) -> * P)
+//          remove \forall P:Int in all above three
+```
+
+where `isPrime(_)` is a predicate that can be defined in the usual way. 
+
+As another example, `randDifferentFromList(Is)` takes a list `Is` of integers and can rewrite to every integers that are not in `Is`. Here is its axiom
+
+```
+(RANDLIST) \forall Is:List{Int} . \forall I:Int . \neg(I \in Is) -> randDifferentFromList(Is) -> * I
+```
+
+### Frontend Syntax
+
+We do not have a frontend syntax proposal for `rand()` and its variants. This is left as future work. 
+
+## Example B: Fresh Integer Construct `fresh(Is)`
+
+The fresh integer construct `fresh(Is)` is a language construct and part of the PL syntax
+
+```
+Stmt ::= ... | "fresh" "(" List{Int} ")"
+```
+
+### Axioms
+
+The intended semantics of `fresh(Is)` is that it can rewrite to any integers that are not in `Is`.  
+
+Note that `fresh(Is)` and `randDifferentFromList(Is)` are different; the former does not need to rewrite to _every_ integers not in `Is`, while the latter requires so. For example, it is _correct_ to implement `fresh(Is)` so it always returns the smallest positive integers that are not in `Is`, but same implementation for `randDifferentFromList(Is)` is considered _incorrect_. In other words, there exist multiple correct implementations of `fresh(Is)`, some of which may be deterministic, but there only exists a unique implementation of `randDifferentFromList(Is)`. Finally, it is a _correct_ if one implements `fresh(Is)` the same as `randDifferentFromList(Is)`. 
```

Comment:
smallest positive integer that is not? the/a set of smallest positive integers?

also: "a correct"?